### PR TITLE
 encapsulate videoSettings into player

### PR DESCRIPTION
**Pull request by FernetMenta**
VideoSettings were abused as globals, writing and reading not threadsafe. Determining what file "CurrentVideoSettings" belong to was error prone and led to saving settings to wrong file. Having VideoPlayer dependent on a singleton made it a singleton as well. In short: all was wrong with videoSettings.

### CODE REVIEW

**Comment by FernetMenta** [comment]
@popcornmix omxplayer may need some adjustments

**Comment by FernetMenta** [comment]
@garbear RetroPlayer followed a bad example and placed its settings into MediaSettings too. I suggest to change this. I will move AudioSettings out of MediaSettings.

File: xbmc/cores/VideoPlayer/Process/ProcessInfo.cpp [hunk 1]
```diff
@@ -600,3 +600,24 @@ int64_t CProcessInfo::GetMaxTime()
...
+CVideoSettings CProcessInfo::GetVideoSettings()
+{
+  CSingleLock lock(m_settingsSection);
+  return m_videoSettings;
+}
+
+CVideoSettings& CProcessInfo::UpdateVideoSettigs()
+{
+  CSingleLock lock(m_settingsSection);
+  return m_videoSettings;
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/Process/ProcessInfo.cpp]: `m_videoSettings` shouldn't be accessed after lock is destructed

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/Process/ProcessInfo.cpp]: I know but that needs an additional round to get fixed. currently there is no lock at all

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 2]
```diff
@@ -1228,6 +1261,10 @@ void CVideoPlayer::Prepare()
   m_CurrentAudio.lastdts = DVD_NOPTS_VALUE;
   m_CurrentVideo.lastdts = DVD_NOPTS_VALUE;
 
+  CJobManager::GetInstance().Submit([&]() {
+    m_callback.RequestVideoSettings(m_item);
```
Comment (garbear) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: I think we addressed before the use of references in a JobManager lambda?

Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp]: thanks! fixed now

File: xbmc/cores/VideoPlayer/Process/ProcessInfo.cpp (hunk 1 above)
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/Process/ProcessInfo.cpp]: update is threadsafe now

**Comment by ace20022** [comment]
I think ```CGUIDialogVideoSettings::OnSettingChanged``` isn't thread-safe. Could you check that while you're at it?

**Comment by FernetMenta** [comment]
Why do you think it is not threadsafe with this PR?

**Comment by ace20022** [comment]
If ```CGUIDialogVideoSettings::OnSettingChanged``` gets called from multiple threads (not sure whether this can really happen or not) then each thread gets a copy ``` CVideoSettings vs```which it manipulates and writes back via ```g_application.m_pPlayer->SetVideoSettings(vs)``` possibly overriding the changes from the other treads.

**Comment by FernetMenta** [comment]
@ace20022 I don't consider this as a threading issue. Even if you just wrote a single value and Kodi allows manipulation of settings via multiple clients, you would have a race, one client overwrites the value of the other.

Changing videoSettings from more than a single client at the same time makes no sense. That's like one client wants to go ff and another one rw.

File: xbmc/cores/VideoPlayer/DVDCodecs/Video/VDPAU.cpp [hunk 3]
```diff
@@ -1875,28 +1874,28 @@ void CMixer::CheckFeatures()
...
     SetNoiseReduction();
   }
-  if (m_Sharpness != CMediaSettings::GetInstance().GetCurrentVideoSettings().m_Sharpness)
+  if (m_Sharpness != m_config.processInfo->GetVideoSettings().m_Sharpness)
   {
-    m_Sharpness = CMediaSettings::GetInstance().GetCurrentVideoSettings().m_Sharpness;
+    m_Sharpness = m_config.processInfo->GetVideoSettings().m_Sharpness;
     SetSharpness();
   }
-  if (m_Deint != CMediaSettings::GetInstance().GetCurrentVideoSettings().m_InterlaceMethod)
```
[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/DVDCodecs/Video/VDPAU.cpp]: SetDeinterlacing is much more than a single operation. You certainly don't want to call this on every cycle.

[the remark this answers is not in the document]
Comment (hauihau) [reply xbmc/cores/VideoPlayer/DVDCodecs/Video/VDPAU.cpp]: Sorry for the noise, I didn't see that there was another function call.

**Comment by ace20022** [comment]
@FernetMenta okay, it just came to my mind while skimming through the pr. The formatting around https://github.com/xbmc/xbmc/pull/12944/files#diff-d920d5ea15baa1b7ba5b617ae9b76489R281 seems to be wrong.

**Comment by popcornmix** [comment]
@FernetMenta the omxplayer changes seem okay, but we handle deinterlace in mmalrenderer:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/MMALRenderer.cpp#L689
Currently renderers don't get processinfo passed in - how should I get m_InterlaceMethod from here?

**Comment by FernetMenta** [comment]
@popcornmix CBaseRenderer has videoSettings: https://github.com/FernetMenta/xbmc/blob/videosettings/xbmc/cores/VideoPlayer/VideoRenderers/BaseRenderer.h#L139

**Comment by FernetMenta** [comment]
@ace20022 I did some polishing on the formatting.

**Comment by FernetMenta** [comment]
jenkins build this please

**Comment by mcaptur** [comment]
@FernetMenta amlogic does not build with this pr

**Comment by FernetMenta** [comment]
@peak3d can you fix this?

**Comment by FernetMenta** [comment]
@mcaptur I noticed that you fixed it already in our repo. why don't you send this patch upstream?

**Comment by peak3d** [comment]
I'll wait for @mcaptur 's resonse....

**Comment by mcaptur** [comment]
no tried a quick patch but didn't work and am away at the moment.. if no one fixes it i should have some time coming week.

**Comment by peak3d** [comment]
@FernetMenta this PR is merged, and I just build mainline master without issues. Sry, I don't see the problem.

**Comment by FernetMenta** [comment]
you built aml without issues?

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 4]
```diff
@@ -3217,6 +3253,7 @@ float CVideoPlayer::GetAVDelay()
 
 void CVideoPlayer::SetSubTitleDelay(float fValue)
 {
+  m_processInfo->UpdateVideoSettigs().SetAudioDelay(fValue);
```
Comment (wiromare) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: small typo, it should be SetSubtitleDelay

@FernetMenta

**Comment by notspiff** [comment]
And VideoSettiNgs

28. okt. 2017 23:00 skrev "wiromare" <notifications@github.com>:

*@wiromare* commented on this pull request.
------------------------------

In xbmc/cores/VideoPlayer/VideoPlayer.cpp
<https://github.com/xbmc/xbmc/pull/12944#discussion_r147563410>:

> @@ -3217,6 +3253,7 @@ float CVideoPlayer::GetAVDelay()

 void CVideoPlayer::SetSubTitleDelay(float fValue)
 {
+  m_processInfo->UpdateVideoSettigs().SetAudioDelay(fValue);

small typo, it should be SetSubtitleDelay

@FernetMenta <https://github.com/fernetmenta>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/xbmc/xbmc/pull/12944#pullrequestreview-72680554>, or mute
the thread

File: xbmc/video/PlayerController.cpp [hunk 5]
```diff
@@ -132,67 +132,71 @@ bool CPlayerController::OnAction(const CAction &action)
...
       case ACTION_AUDIO_DELAY_MIN:
       {
-        CMediaSettings::GetInstance().GetCurrentVideoSettings().m_AudioDelay -= 0.025f;
-        if (CMediaSettings::GetInstance().GetCurrentVideoSettings().m_AudioDelay < -g_advancedSettings.m_videoAudioDelayRange)
-          CMediaSettings::GetInstance().GetCurrentVideoSettings().m_AudioDelay = -g_advancedSettings.m_videoAudioDelayRange;
-        g_application.m_pPlayer->SetAVDelay(CMediaSettings::GetInstance().GetCurrentVideoSettings().m_AudioDelay);
+        CVideoSettings vs = g_application.m_pPlayer->GetVideoSettings();
+        vs.m_AudioDelay -= 0.025f;
+        if (vs.m_AudioDelay < -g_advancedSettings.m_videoAudioDelayRange)
+          vs.m_SubtitleDelay = -g_advancedSettings.m_videoAudioDelayRange;
```
Comment (ace20022) [inline xbmc/video/PlayerController.cpp]: Should be vs.m_AudioDelay

**Pull request merged**